SearchKit - Add new "Tree" display 🌲, use for AdminUI Navigation menu screen

colemanw opened the pull request:
Overview
----------------------------------------
New SearchKit "Tree" display provides similar functionality to our old jQuery.jstree library, allowing more older UI screens to be updated.
This implements it for the Navigation screen in AdminUI:

![image](https://github.com/user-attachments/assets/14ef44e1-988c-44ac-90e9-a85e8448e98c)

ufundo committed:
Api4 - expose Navigation items to search kit

correct parent_field specification for Navigation entity

ufundo committed:
AdminUI - add SearchKit for editing Nav menu

AdminUI - limit nav menu editor to current domain; exclude home

Co-authored-by: colemanw <coleman@civicrm.org>

AdminUI Nav Editor - use action buttons, add weight column, update Enable/Disable styles; fix lint

colemanw committed:
AdminUI - Switch 'Administer Navigation Menu' to new Tree display

Also switches the server_route to fully replace the old screen

colemanw committed:
SearchKit - fix compatability between hierarchical and editable rows

During editable refreshes, the hierarchical state wasn't being recalculated server-side, but it doesn't need to be,
we just needed to move hierarchical style logic from PHP to JavaScript so the state could be preserved.
Calculating default collapsed states client-side means it won't be overwritten during refresh, and since _depth and _children
doesn't change during inline-edit, they can be carried over without the need to recalculate.

colemanw committed:
SearchKit - Use InlineEdit action for draggable sorting

Improves efficiency and permission handling by going through the new InlineEdit action
instead of the vanilla api4 to update weights.

colemanw committed:
SearchKit - Add support for updating parent_field in inline edits

Previously, inline editing did not allow updating parent_field for hierarchical entities. This change ensures that parent_field values are included when applicable. It enhances flexibility for managing hierarchical data through the interface.

colemanw committed:
SearchKit - Use row key instead of row index for inline editing

This works better with non-flat data structures, e.g. tree displays

colemanw committed:
SearchKit - Improve performance of clientside field rendering

While testing the new AdminUI "Administer Navigation Menu" tree display, which has hundreds
of rows, performance became a problem.

This improves the AngularJS template efficiency by avoiding function watchers and using
one-time binding as much as possible.

The field.html template has been split into 3, to make each one simpler and more efficient.

colemanw committed:
SearchKit - Add new SearchDisplay type "Tree"

Introduces a new tree search display type, enabling hierarchical data representation in the search results. Features include collapsible branches and drag-n-drop sorting.

This should allow us to replace the old jQuery.jsTree library with SearchKit.

colemanw commented:
@vingle this PR adds a new css file to core. What's your preferred way to communicate/collaborate whenever css changes are made to core?

vingle commented:
@colemanw - funny was just updating some changes from /core/css today _and_ looking at JSTree. I guess just a ping is good - then I can at least test, and potentially mirror the file directly to swap variables in (e.g. for tree-placeholder).

Nice to see a JStree replacement btw!

aydun commented:
Looks great!  A few things:
1) It uses  civicrm/admin/navigation not civicrm/admin/menu so the default Admin > Customise > Navigation Menu uses the old version
2) In-place editing menu labels has the save/cancel icons in the wrong place:
![image](https://github.com/user-attachments/assets/23c342a2-8175-48bf-9b54-537fb14834ed)
3) Minor: if you drag an item between menus then it seems to retain its old position in the list, not the new.  For example, drag User Guide from Support to the bottom of Administer and then reload the page.  User Guide then shows as the second item under Administer, not at the bottom.   If you then drag it to the bottom of Administer again, it then saves correctly.

colemanw commented:
Thanks for testing this @aydun! What theme are you using? The buttons do align for me, using the standard Greenwich:
![image](https://github.com/user-attachments/assets/111f0539-97d1-4605-8087-eb09167bc302)

aydun commented:
@colemanw That screenshot was with the Greenwich default on the build test site  core-31803-6fpee.test-1.civicrm.org:8003 

With RiverLea/Minetta, it is very messed up:
![image](https://github.com/user-attachments/assets/5862dedf-0c27-4a1b-aa01-c1fed8d1ee9b)

vingle commented:
@aydun I noticed.. I think that's because none of the new CSS here is integrated into RL yet - will wait for this to get merged first, unless that's going to take a while.

colemanw commented:
Thanks @aydun I reproduced it on Firefox as well, but it's a pre-existing issue (also affects search displays of type _List_) so let's not hold up this PR on that. I'll look into the sort ordering bug tho.

colemanw committed:
Api4 - Fix weight calculation when moving items to/from a parent group

Fixes bugs in the drag-n-drop functionality of the SearchKit Tree displays.

Before: Weights not adjusted when moving item into a new group;
update fails when moving item to the top-level of the tree.

After: Weights adjust correctly whether moving within or between branches of the tree.

The weight handling code needed updating to not conflate NULL with 0.

colemanw commented:
@aydun I pushed up a fix to the sort-order issue + a unit test, and also tweaked the javascript & css to make the drag-n-drop a bit easier to use. Ready for a final review.

colemanw committed:
AdminUI - Add separator styles to navigation menu editor

Visually shows the menu item separators in the tree

aydun commented:
The most recent test site didn't build - redirects to http://core-31803-6jkax.test-1.civicrm.org:8003/install.php

Test this please.

colemanw commented:
@aydun it appears to be a problem with our demo/CI server, being discussed on https://chat.civicrm.org/civicrm/channels/infrastructure

colemanw commented:
@aydun - it looks like @totten fixed the problem. New demo site successfully built.

aydun commented:
@colemanw That seems to have fixed the drag and drop ordering.

Something has broken with in-place editing - click, make the change, hit the green tick ... and you get a pulsating grey box.  If you reload the page, the new version is shown so it is being saved but not refreshing correctly.

Enabling/Disabling works but does not refresh so the disabled icon and light grey text do not show until the page is reloaded.

FWIW, browser console shows 'Error: Invalid tab ID: 619' on loading any page.

And the page route is still different to the menu entry Navigation Menu.

colemanw commented:
@aydun the problem was the newest commit in this PR: https://github.com/colemanw/civicrm-core/commit/a42282c0a12d5b2c965867d9ab99fba0867676fb . I'm just gonna delete that commit from this PR and open a followup for it once this is merged.
Done. And I also just updated the route so this screen now fully replaces the old one. Hey as long as we're doing this, might as well fully commit eh?

aydun commented:
Mostly dragging items between parent menus works, but there seems to be a an edge case if you drag the item to be the last entry in a different parent menu - in that case, it retains its existing weight.

But that is rare and easily worked around so let's merge and you can followup with a fix if you want.

Another notable step forward!

ufundo commented:
Awesome work, thanks for picking this up @colemanw and reviewing @aydun !

colemanw commented:
Thanks @aydun - would you be able to merge this follow-up? https://github.com/civicrm/civicrm-core/pull/31866
